test/test_models.py

```diff
@@ -69,16 +94,269 @@ def check_script(self, model, name):
             msg = str(e) + str(tb)
         self.assertTrue(scriptable, msg)
 
-    def _test_classification_model(self, name, input_shape):
-        # passing num_class equal to a number other than 1000 helps in making the test
-        # more enforcing in nature
-        model = models.__dict__[name](num_classes=50)
-        self.check_script(model, name)
-        model.eval()
-        x = torch.rand(input_shape)
-        out = model(x)
-        self.assertEqual(out.shape[-1], 50)
+    def _check_scriptable(self, model, expected):
+        if expected is None: # we don't check scriptability for all models
+            return
+
+        actual = True
+        msg = ''
+        try:
+            torch.jit.script(model)
+        except Exception as e:
+            tb = traceback.format_exc()
+            actual = False
+            msg = str(e) + str(tb)
+        self.assertEqual(actual, expected, msg)
+
+
+
+class ClassificationCoverageTester(TestCase):
+
+    # Find all models exposed by torchvision.models factory methods (with assumptions)
+    def get_available_classification_models(self):
+        # TODO add a registration mechanism to torchvision.models
+        return [k for k, v in models.__dict__.items() if callable(v) and k[0].lower() == k[0] and k[0] != "_"]
+
+    # Recursively gather test methods from all classification testers
+    def get_test_methods_for_class(self, klass):
+        all_methods = inspect.getmembers(klass, predicate=inspect.isfunction)
+        test_methods = set([method[0] for method in all_methods if method[0].startswith('test_')])
+        for child in klass.__subclasses__():
+            test_methods = test_methods.union(self.get_test_methods_for_class(child))
+        return test_methods
+
+    # Verify that all models exposed by torchvision.models factory methods
+    #    have corresponding test methods
+    # NOTE This does not include some of the extra tests (such as Resnet
```

Comment:
This does not enforce that all tests exposed are tested, only that names of the models are present in the test names.

So if someone just copy-pastes a test and modify the test name, without modifying the body of the test, the test would still fail.
Here is an example

```python
    def test_vgg19(self):
        self._test_classification_model(models.vgg19)

    def test_vgg19_bn(self):
        self._test_classification_model(models.vgg19)
```
I think this is pretty hard to spot for a reviewer, and tests would pass without problems. 
For me, this hints that we should avoid having manual test writing for the model, and that the previous approach (which iterated over the models) was safer.

Reply:
If the problem is that we can't keep people from doing obviously incorrect things, I agree that's a problem, but I don't think we're going to fix that problem with code.

Reply:
The previous dynamic method of constructing tests also suffers from this issue. Adding a model would automatically add a test _without checking that the dynamically added test was correct for the new model._ This is also a violation of the  principle that tests should be intentional.

The new state of affairs is an improvement: Adding a model does _not_ automatically add a test, and in fact causes a failure and demands that the dev add an appropriately named test. It doesn't take the place of a proper test coverage tool, but acts as a nudge to get the developer to add a test that complies with the naming convention.

Reply:
I think there are two points here worth discussion:
1 - making sure that the code runs without errors for the environments we support
2 - verifying that they give the same result for a given model on all systems

I was mentioning point number 1 in my first comment.
In the current approach, a model can be entirely not tested (execution test) because of a bad copy-paste like the one I mentioned. In that situation, the code satisfies the "coverage" test, given that all models are covered by a corresponding function name, but the test named `resnet50` can run `resnet18` internally.

About point 2, we can also make it a hard failure to add a new model without adding a corresponding correctness model file. All the user would need to do is to run the test code with the `--accept` flag to generate a new test file.

So I believe we can have both coverage and correctness with the autogenerated files, without requiring the user to add extra (repeated) lines in the code.

Let me know if I'm missing something

Reply:
I've discussed with @lerks about this, and he took [the Zen of Python](https://www.python.org/dev/peps/pep-0020/) as an example:

> Explicit is better than implicit.

In this sense, an autogenerated test is less desirable because it's harder to grep for.

Given that, I'm resolving this conversation.